django/db/backends/mysql/operations.py

```diff
@@ -177,13 +177,45 @@ def sql_flush(self, style, tables, sequences, allow_cascade=False):
         # NB: The generated SQL below is specific to MySQL
         # 'TRUNCATE x;', 'TRUNCATE y;', 'TRUNCATE z;'... style SQL statements
         # to clear all tables of all data
+        schema_name = self.connection.settings_dict['NAME']
+
         if tables:
+            with self.connection.cursor() as cursor:
+                cursor.execute("""
+                SELECT table_name, table_rows, auto_increment
+                FROM information_schema.tables
+                WHERE table_schema = %s;
+                """, (schema_name,))
+                rows = cursor.fetchall()
+
             sql = ['SET FOREIGN_KEY_CHECKS = 0;']
-            for table in tables:
-                sql.append('%s %s;' % (
-                    style.SQL_KEYWORD('TRUNCATE'),
-                    style.SQL_FIELD(self.quote_name(table)),
-                ))
+
+            for table_name, table_rows, auto_increment in rows:
+                if table_name not in tables:
+                    continue
+
+                if table_rows <= 1000:
+                    # 1000 is a heuristic threshold to determine
+                    # whether using "DELETE FROM" or "TRUNCATE".
+                    sql.append('%s %s %s;' % (
+                        style.SQL_KEYWORD('DELETE'),
+                        style.SQL_KEYWORD('FROM'),
+                        style.SQL_FIELD(self.quote_name(table_name)),
+                    ))
+
+                    if auto_increment is not None:
+                        sql.append('%s %s %s %s = 1;' % (
+                            style.SQL_KEYWORD('ALTER'),
+                            style.SQL_KEYWORD('TABLE'),
+                            style.SQL_FIELD(self.quote_name(table_name)),
+                            style.SQL_FIELD('AUTO_INCREMENT'),
+                        ))
```

Comment:
Is this necessary to get tests passing? I think it's already handled by `self.sequence_reset_by_name_sql(style, sequences)` below when `sequences` is specified.
```suggestion
```

Tests that require sequences to be reset should explicitly set [`reset_sequences=True`](https://docs.djangoproject.com/en/3.0/topics/testing/advanced/#django.test.TransactionTestCase.reset_sequences).

Reply:
Hi. Thank you for your review.

I applied your suggestion at https://github.com/django/django/pull/12634/commits/df2177682ad13bc5b5660ac302b6eaa9e6ab86ce, but tests are failed. It seems to be required. Can I revert the change?

Reply:
Tests are failing for a different reason due to qwirks with our CI system; notice that SQLite tests are failing as well.

Reply:
Oh, actually it looks like `test_runner.tests.AutoIncrementResetTest.test_autoincrement_reset2` is also failing on MySQL.

Reply:
There's seem to be something else off here, the `reset_sequence` code should take care of resetting sequences appropriately on this test

https://github.com/django/django/blob/b9336b78cf2a9a29f4934041c9e221bc68daec80/django/test/testcases.py#L961-L970

Reply:
`sequence_reset_by_name_sql` is not defined in `mysql.operations.DatabaseOperations`. So it seems to call following method in `BaseDataBaseOperations`.

https://github.com/django/django/blob/227d0c7365cfd0a64d021cb9bdcf77bed2d3f170/django/db/backends/base/operations.py#L408-L416

Reply:
I implemented it at 93531bb, then tests are passed.